compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java

```diff
@@ -1079,14 +1103,19 @@ public void visit(BLangTupleTypeNode tupleTypeNode) {
         }
 
         BTypeSymbol tupleTypeSymbol = Symbols.createTypeSymbol(SymTag.TUPLE_TYPE, Flags.asMask(EnumSet.of(Flag.PUBLIC)),
-                Names.EMPTY, env.enclPkg.symbol.pkgID, null, env.scope.owner);
+                                                               Names.EMPTY, env.enclPkg.symbol.pkgID, null,
+                                                               env.scope.owner);
 
+        List<BType> allMemTypes = new ArrayList<>(memberTypes);
         BTupleType tupleType = new BTupleType(tupleTypeSymbol, memberTypes);
         tupleTypeSymbol.type = tupleType;
-        if (tupleTypeNode.restParamType !=  null) {
+        if (tupleTypeNode.restParamType != null) {
             tupleType.restType = resolveTypeNode(tupleTypeNode.restParamType, env);
+            allMemTypes.add(tupleType.restType);
         }
 
+        markParameterizedType(tupleType, allMemTypes);
```

Comment:
we can just make this `markParameterizedType(tupleType, memberTypes)` and call markParameterizedType again when `tupleTypeNode.restParamType != null` with `restParamType`
`markParameterizedType(tupleType, tupleTypeNode.restParamType)`.
With caching on `markParameterizedType` we can avoid creating a new `ArrayList` completely.

Reply:
Ah if changing it like this, your previous suggestions to check if it's a parameterized type before looping makes sense. Will change